Add support for lcs-lead (like lcs-trail but for leading spaces)

benoncoffee opened the pull request:
I wanted to have a listchars option for leading spaces instead of for trailing spaces (it seems like there are more people interested in this, see [this question on stackoverflow](https://stackoverflow.com/questions/4617059/showing-trailing-spaces-in-vim) for example). This didn't exist according to the very nice people on the #vim channel on IRC, so I implemented it.

I mostly mimicked the code for lcs_trail. My understanding of the codebase is not perfect, so I might have made errors (I am not sure about the changes in screen.c, although everything seems to build without errors, "make test" gives "ALL DONE", and the implemented functionality seems to work well).

Commit:
Add support for lcs-lead (like lcs-trail but for leading spaces)

dpelle commented on runtime/doc/options.txt:
```diff
@@ -5007,6 +5007,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 			trailing spaces are blank.  Overrides the "space"
 			setting for trailing spaces.
+							*lcs-lead*
+	  lead.c	Character to show for leading spaces. When omitted,
```
It looks like it should be lead:c instead of lead.c

cvwillegen reviewed (changes requested):
_leading_ spaces are blank.

cvwillegen commented on runtime/doc/options.txt:
```diff
@@ -5007,6 +5007,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 			setting for trailing spaces.
+							*lcs-lead*
+	  lead:c	Character to show for leading spaces. When omitted,
+			trailing spaces are blank. Overrides the "space"
```
_leading_ spaces are blank

cvwillegen commented on src/message.c:
```diff
@@ -1896,6 +1904,11 @@ msg_prt_line(char_u *s, int list)
 		c = lcs_trail;
 		attr = HL_ATTR(HLF_8);
 	    }
+	    else if (c == ' ' && lead != NULL && s < lead)
```
On these lines, if a line is full of spaces, they are all _trailing spaces_, below (see other comment...)

cvwillegen commented on src/screen.c:
```diff
@@ -4828,6 +4836,28 @@ win_line(
+#endif
+		}
+
+		if (leadcol != 0 && ptr <= line + leadcol && c == ' ')
```
Here, it looks like a line full of spaces is seen as leading spaces. Is this as intended?

It would be wise to write tests for these changes, that way you would see if your (or my!) assumptions about code are correct or incorrect...

brammool committed:
patch 8.1.0153: build with SHADOWDIR fails

Problem:    Build with SHADOWDIR fails. (Elimar Riesebieter)
Solution:   Create a link for Make_all.mak. (Tony Mechelynck)

brammool commented:
> > @@ -4828,6 +4836,28 @@ win_line(
>  		    }
>  		    else
>  			mb_utf8 = FALSE;
> +#endif
> +		}
> +
> +		if (leadcol != 0 && ptr <= line + leadcol && c == ' ')
> 
> That is a really good observation. It raises the question: Should
> (assuming that lcs_trail and lcs_lead are both set) a line of spaces
> be treated as *trailing* or *leading* spaces?
> 
> I didn't really think about this. In all honesty, I don't see a reason
> to prefer one over the other.

A line with only whitespace has trailing spaces.  I see leading spaces
as spaces that go before a non-space character.  While trailing spaces
can be after anything, including nothing.

-- 
hundred-and-one symptoms of being an internet addict:
189. You put your e-mail address in the upper left-hand corner of envelopes.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

brammool committed:
patch 8.1.0154: crash with "set smarttab shiftwidth=0 softtabstop=-1"

Problem:    Crash with "set smarttab shiftwidth=0 softtabstop=-1".
Solution:   Fall back to using 'tabstop'. (closes #3155)

brammool committed:
patch 8.1.0155: evim.man missing from the distribution

Problem:    Evim.man missing from the distribution.
Solution:   Add it to the list.

brammool committed:
patch 8.1.0156: MS-Windows compiler warning

Problem:    MS-Windows compiler warning.
Solution:   Add a type cast. (Mike Williams)

brammool committed:
patch 8.1.0157: old iTerm2 is not recognized, resulting in stray output

Problem:    Old iTerm2 is not recognized, resulting in stray output.
Solution:   Recognize the termresponse.

brammool committed:
patch 8.1.0158: GUI: input() fails if CTRL-C was pressed before

Problem:    GUI: input() fails if CTRL-C was pressed before. (Michael Naumann)
Solution:   call vpeekc() to drop the CTRL-C from the input stream.

brammool committed:
patch 8.1.0159: completion for user names does not work for a prefix.

Problem:    Completion for user names does not work if a prefix is also a full
            matching name. (Nazri Ramliy)
Solution:   Accept both full and partial matches. (Dominique Pelle)

brammool committed:
patch 8.1.0160: no Danish manual translations

Problem:    No Danish manual translations.
Solution:   Add the Danish manual translations to the file list.

brammool committed:
patch 8.1.0161: buffer not updated with 'autoread' set if file was deleted

Problem:    Buffer not updated with 'autoread' set if file was deleted.
            (Michael Naumann)
Solution:   Don't set the timestamp to zero. (closes #3165)

brammool committed:
patch 8.1.0162: Danish and German man pages are not installed

Problem:    Danish and German man pages are not installed. (Tony Mechelynck)
Solution:   Adjust the makefile

brammool committed:
patch 8.1.0163: insufficient testing for Tcl

Problem:    Insufficient testing for Tcl.
Solution:   Add a few more tests. (Dominique Pelle, closes #3166)

brammool committed:
patch 8.1.0164: luaeval('vim.buffer().name') returns an error

Problem:    luaeval('vim.buffer().name') returns an error.
Solution:   Return an empty string. (Dominique Pelle, closes #3167)

brammool committed:
Update runtime files, add Danish translations.

brammool committed:
patch 8.1.0165: :clist output can be very long

Problem:    :clist output can be very long.
Solution:   Support filtering :clist entries. (Yegappan Lakshmanan)

brammool committed:
patch 8.1.0166: using dict_add_nr_str() is clumsy

Problem:    Using dict_add_nr_str() is clumsy.
Solution:   Split into two functions. (Ozaki Kiichi, closes #3154)

brammool committed:
patch 8.1.0167: lock flag in new dictitem is reset in many places

Problem:    Lock flag in new dictitem is reset in many places.
Solution:   Always reset the lock flag.

brammool committed:
patch 8.1.0168: output of :marks is too short with multi-byte chars

Problem:    Output of :marks is too short with multi-byte chars. (Tony
            Mechelynck)
Solution:   Get more bytes from the text line.

brammool committed:
patch 8.1.0169: calling message_filtered() a bit too often

Problem:    Calling message_filtered() a bit too often.
Solution:   Only call message_filtered() when filtering is already false.

brammool committed:
patch 8.1.0170: invalid memory use with complicated pattern

Problem:    Invalid memory use with complicated pattern. (Andy Massimino)
Solution:   Reallocate the list of listids when needed. (closes #3175)
            Remove unnecessary function prototypes.

brammool committed:
patch 8.1.0171: typing CTRL-W n in a terminal window causes ml_get error

Problem:    Typing CTRL-W n in a terminal window causes ml_get error.
Solution:   When resizing the terminal outside of terminal_loop() make sure
            the snapshot is complete.

brammool committed:
patch 8.1.0172: 'viminfofile' option does not behave like a file name

Problem:    'viminfofile' option does not behave like a file name.
Solution:   Add the P_EXPAND flag. (closes #3178)

brammool committed:
patch 8.1.0173: compiler warning on MS-Windows

Problem:    Compiler warning on MS-Windows.
Solution:   Add type cast. (Mike Williams)

brammool committed:
patch 8.1.0174: after paging up and down fold line is wrong

Problem:    After paging up and down fold line is wrong.
Solution:   Correct the computation of w_topline and w_botline. (Hirohito
            Higashi)

brammool committed:
patch 8.1.0175: marks test fails in very wide window

Problem:    Marks test fails in very wide window. (Vladimir Lomov)
Solution:   Extend the text to match 'columns'. (closes #3180, closes #3181)

brammool committed:
patch 8.1.0176: overlapping string argument for strcpy()

Problem:    Overlapping string argument for strcpy(). (Coverity)
Solution:   Use STRMOVE() instead of STRCPY(). (Dominique Pelle, closes #3187)

brammool committed:
patch 8.1.0177: defining function in sandbox is inconsistent

Problem:    Defining function in sandbox is inconsistent, cannot use :function
            but can define a lambda.
Solution:   Allow defining a function in the sandbox, but also use the sandbox
            when executing it. (closes #3182)

brammool committed:
patch 8.1.0178: warning for passing pointer to non-pointer argument

Problem:    Warning for passing pointer to non-pointer argument.
Solution:   Use zero instead of NULL.

brammool committed:
patch 8.1.0179: redundant condition for boundary check

Problem:    Redundant condition for boundary check.
Solution:   Remove the condition. (Dominique Pelle). Change FALSE to FAIL.

chrisbra commented:
Is there anything, that cannot achieved by using `machadd()` in combination with concealing?

brammool committed:
patch 8.1.0180: static analysis errors in Lua interface

Problem:    Static analysis errors in Lua interface. (Coverity)
Solution:   Check for NULL pointers.

brammool committed:
patch 8.1.0181: memory leak with trailing characters in skip expression

Problem:    Memory leak with trailing characters in skip expression.
Solution:   Free the return value.

brammool committed:
patch 8.1.0182: Unicode standard was updated

Problem:    Unicode standard was updated.
Solution:   Include the changes. (Christian Brabandt)

brammool committed:
patch 8.1.0183: Lua API changed, breaking the build

Problem:    Lua API changed, breaking the build.
Solution:   Adjust prototype of lua_rawgeti(). (Ken Takata,
            closes #3157, closes #3144)

brammool committed:
patch 8.1.0184: not easy to figure out the window layout

Problem:    Not easy to figure out the window layout.
Solution:   Add "wincol" and "winrow" to what getwininfo() returns.

brammool committed:
patch 8.1.0185: running tests writes lua.vim even though it is not used

Problem:    Running tests writes lua.vim even though it is not used.
Solution:   Stop writing lua.vim.

brammool committed:
patch 8.1.0186: test for getwininfo() fails in GUI

Problem:    Test for getwininfo() fails in GUI.
Solution:   Account for missing tabline.

Commit:
Merge branch 'master' of https://github.com/vim/vim

brammool commented:
It seems comments were not responded to and a test is lacking.  Therefore closing.  If someone still wants this, please open a new PR and take care of the comments.
